plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,504 @@
+              pure $ case accTerm of
+                       -- MERGE current let-group with previous let-group iff both groups' recursivity is NonRec
+                       Let accAnn NonRec accBs accIn | mRecurs == NonRec -> Let (mAnn <> accAnn) NonRec (newBindings <> accBs) accIn
+                       -- never merge if the previous let-group is a Rec or the current let-group is Rec,
+                       -- but instead create a nested current let-group under the previous let-group (above)
+                       _ -> Let mAnn mRecurs newBindings accTerm
+
+
+-- | The main transformation function (Term -> Term) that returns a term with all lets floated as outwards as possible (full laziness).
+-- Is comprised of two AST "passes":
+-- 1stpass: to collect the ranks (positions) of all lets
+-- 2ndpass:  to remove all its lets and place them back (float them) in their ranks (positions).
+-- See Note [Float algorithm]
+--
+-- NB: This transformation requires that the PLC.rename compiler-pass has prior been run.
+floatTerm :: forall name tyname uni a
+           . (PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique, Monoid a)
+          => Term tyname name uni a -> Term tyname name uni a
+floatTerm pir = p2Term pir
+                -- give the floatdata of the 1st pass to the start of the 2nd pass
+              $ p1Term pir
+
+
+
+-- Helpers
+----------
+
+-- | A getter that returns a single 'Unique' for a particular binding.
+-- We need this because let-datatypes introduce multiple identifiers, but in our 'RhsTable', we use a single Unique as the key. See also: 'bindingIds'
+principal :: (PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
+            => Getting r (Binding tyname name uni a) PLC.Unique
+principal = to $ \case TermBind _ _ (VarDecl _ n _) _ -> n ^. PLC.theUnique
+                       TypeBind _ (TyVarDecl _ n _) _ -> n ^. PLC.theUnique
+                       -- arbitrary: uses the type construtors' unique as the principal unique of this data binding group
+                       DatatypeBind _ (Datatype _ (TyVarDecl _ tConstr _) _ _ _) -> tConstr ^. PLC.theUnique
+
+
+-- | given the lets to float here, it splits the SCC state to a pair of:
+-- 1) sccs to process at this location
+-- 2) and remainder sccs to use as the next state.
+splitSccs :: Ord a
+          => S.Set a -> [S.Set a] -> ([S.Set a], [S.Set a])
+splitSccs lets =
+  let initAcc = (lets, (mempty,mempty))
+  in snd . foldr (\ scc (ls, (hs,rs)) ->
+              let commonScc = scc `S.intersection` ls
+                  remLets = lets S.\\ commonScc
+                  remScc = scc S.\\ commonScc
+              in (remLets, ([commonScc | not $ null commonScc] <> hs -- skip appending empty sccs
+                           ,[remScc | not $ null remScc] <> rs) -- skip appending empty sccs
+                 )
+           ) initAcc
+
+isSingleton :: Foldable t => t a -> Bool
```

Comment:
I can't decide if this function is terrible or brilliant :laughing:

Reply:
It's a little unfortunate that it needs to traverse the entire container just to check if it contains only a single element. We could use `foldr` and make it lazier, but I guess we don't care.

Reply:
It's both terrible and brilliant :interrobang: 

the default length is foldl' indeed, but it is used here only for `Set`s which have implementation of `length = S.size`. O(1) I will change this to S.size==1 for now.